### ISSUE TITLE
[clang] -Wshadow option produces a warning when explicitly capturing class member using the same name

### ISSUE THREAD

**Issue by flotsamarch** [llvm/llvm-project#71976]
Starting from Clang 17.0.0 (built from source, but this also reproduces on Compiler Explorer) using -Wshadow flag now produces diagnostics for this code:
```
struct A
{
    int b = 5;

    int foo()
    {
        return [b = b] () 
        {
            return b;
        } ();
    }
};
```

I couldn't find this change in changelogs and there is no test for this type of capture in [warn-shadow-in-lambdas.cpp](https://github.com/llvm/llvm-project/blob/main/clang/test/SemaCXX/warn-shadow-in-lambdas.cpp), so I think this is a bug that was introduced by the [commit](https://github.com/llvm/llvm-project/commit/0fb84bc7fdec1c20c2461e54d5c994939fe5b47f) that fixes #61105.

While technically this might be the case of 'shadowing' (I can't make any references to standard that either support or contradict my opinion), I don't think it has ever been considered a 'shadowing' case which should produce diagnostics and has been a somewhat common C++ pattern.

I don't see much value in that warning, since the name `b` referring to original class member is not available inside lambda body, and and the name `b` referring to the lambda member becomes available only when the other `b` is already not. Should it be considered shadowing?

Expected behavior: No warning generated
Actual behavior: `warning: declaration shadows a field of 'A' [-Wshadow]` diagnostics produced

**Comment by myrrc** [comment]
This also reproduces with me on clang 15 and 16 with a slightly different test case:

```cpp
struct A {
  int a;
  void foo() {
    auto b = [a = this->a] {
 
    };
  }
};
```

In this scenario, `a` isn't available in lambda without capturing `this` so there is no real shadowing. Unfortunately, our project https://github.com/ClickHouse/ClickHouse/ is also built with `-Werror` so workarounds like `[a1 = this->a]` have to be written

**Comment by shafik** [comment]
CC @Fznamznon

**Comment by Rinn** [comment]
We're also encountering this in the UnrealEngine codebase as many of our compiles are done with -Wshadow and this has caused hundreds of new warnings. At the very least if this could be put on a separate warning flag that would be really appreciated, for the moment this will block us from updating clang as we don't want to disable -Wshadow entirely to work around it.

**Comment by Fznamznon** [comment]
I don't think it is https://github.com/llvm/llvm-project/commit/0fb84bc7fdec1c20c2461e54d5c994939fe5b47f that is guilty since it adds an orthogonal error message, but I can take a look anyway.

**Comment by cor3ntin** [comment]
`CheckShadow` fails to filter out fields for captures.

```diff
--- a/clang/lib/Sema/SemaDecl.cpp
+++ b/clang/lib/Sema/SemaDecl.cpp
@@ -8351,10 +8351,12 @@ void Sema::CheckShadow(NamedDecl *D, NamedDecl *ShadowedDecl,
 
   unsigned WarningDiag = diag::warn_decl_shadow;
   SourceLocation CaptureLoc;
-  if (isa<VarDecl>(D) && isa<VarDecl>(ShadowedDecl) && NewDC &&
+  if (isa<VarDecl>(D) && NewDC &&
       isa<CXXMethodDecl>(NewDC)) {
     if (const auto *RD = dyn_cast<CXXRecordDecl>(NewDC->getParent())) {
       if (RD->isLambda() && OldDC->Encloses(NewDC->getLexicalParent())) {
+        if(!isa<VarDecl>(ShadowedDecl))
+            return;
         if (RD->getLambdaCaptureDefault() == LCD_None) {
           // Try to avoid warnings for lambdas with an explicit capture list.
           const auto *LSI = cast<LambdaScopeInfo>(getCurFunction());
```

I was able to silence the warning with something along the lines of that.
I am not sure I will have time to submit a PR today so feel free to look more into it if you have time
